agent/consul/state/catalog.go

```diff
@@ -1942,10 +1950,52 @@ func (s *Store) CheckConnectServiceNodes(ws memdb.WatchSet, serviceName string,
 	return s.checkServiceNodes(ws, serviceName, true, entMeta)
 }
 
+// CheckIngressServiceNodes is used to query all nodes and checks for ingress
+// endpoints for a given service.
+func (s *Store) CheckIngressServiceNodes(ws memdb.WatchSet, serviceName string, entMeta *structs.EnterpriseMeta) (uint64, structs.CheckServiceNodes, error) {
+	tx := s.db.Txn(false)
+	defer tx.Abort()
+	maxIdx, nodes, watchChs, err := s.serviceGatewayNodes(tx, serviceName, structs.ServiceKindIngressGateway, entMeta)
+	if err != nil {
+		return 0, nil, fmt.Errorf("failed gateway nodes lookup: %v", err)
+	}
+
+	// TODO(ingress) : Deal with incorporating index from mapping table
+
+	// Watch list of gateway nodes for changes
+	for _, ch := range watchChs {
+		ws.Add(ch)
+	}
+
+	// De-dup service names to lookup
+	serviceNames := make(map[string]struct{})
+	for _, n := range nodes {
+		serviceNames[n.ServiceName] = struct{}{}
+	}
+
+	var results structs.CheckServiceNodes
```

Comment:
Similarly, does this need to be a `map[string]CheckServiceNodes` for namespace stuff?

Reply:
@kyhavlov and I aren't quite sure why `map[string]CheckServiceNodes` would be useful in this case, can you elaborate?

From our understanding, the `NodeService` value of `CheckServiceNodes` has ent metadata on it, and we are only appending values to the results variable here, so nothing would be overwritten.